runtime/shared_common/CompositeCache.cpp

```diff
@@ -1396,32 +1381,61 @@ SH_CompositeCacheImpl::startup(J9VMThread* currentThread, J9SharedClassPreinitCo
+					if ((0 == _theca->roundedPagesFlag)
+						|| _readOnlyOSCache
+					){
 					/* If we don't have rounded pages or if we're running readonly, we can't do mprotect or msync */
-#if defined (J9SHR_MSYNC_SUPPORT)
-					*_runtimeFlags &= ~(J9SHR_RUNTIMEFLAG_ENABLE_MSYNC
+#if defined(J9SHR_MSYNC_SUPPORT)
+						*_runtimeFlags &= ~(J9SHR_RUNTIMEFLAG_ENABLE_MSYNC
 										| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT
 										| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT_ALL
 										| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT_RW
 										| J9SHR_RUNTIMEFLAG_MPROTECT_PARTIAL_PAGES_ON_STARTUP
 										| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT_PARTIAL_PAGES
 										| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT_ONFIND);
-#else
-					*_runtimeFlags &= ~(J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT
+						if (!_readOnlyOSCache) {
+							CC_TRACE(J9SHR_VERBOSEFLAG_ENABLE_VERBOSE, J9NLS_INFO, J9NLS_CC_PAGE_PROTECTION_UNSUPPORTED);
+							CC_TRACE(J9SHR_VERBOSEFLAG_ENABLE_VERBOSE, J9NLS_INFO, J9NLS_CC_MSYNC_UNSUPPORTED);
+						}
+					} else {
+						if (J9_ARE_ALL_BITS_SET(*_runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_PERSISTENT_CACHE)) {
+							int32_t mmapCapabilities = j9mmap_capabilities();
+							if (J9_ARE_NO_BITS_SET(mmapCapabilities, J9PORT_MMAP_CAPABILITY_MSYNC)) {
+								*_runtimeFlags &= ~J9SHR_RUNTIMEFLAG_ENABLE_MSYNC;
+								if (!_readOnlyOSCache) {
+									CC_TRACE(J9SHR_VERBOSEFLAG_ENABLE_VERBOSE, J9NLS_INFO, J9NLS_CC_MSYNC_UNSUPPORTED);
+								}
+							}
+							if (J9_ARE_NO_BITS_SET(mmapCapabilities, J9PORT_MMAP_CAPABILITY_PROTECT)) {
+								*_runtimeFlags &= ~(J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT
+												| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT_ALL
+												| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT_RW
+												| J9SHR_RUNTIMEFLAG_MPROTECT_PARTIAL_PAGES_ON_STARTUP
+												| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT_PARTIAL_PAGES
+												| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT_ONFIND);
+								if (!_readOnlyOSCache) {
+									CC_TRACE(J9SHR_VERBOSEFLAG_ENABLE_VERBOSE, J9NLS_INFO, J9NLS_CC_PAGE_PROTECTION_UNSUPPORTED);
+								}
+							}
+						}
+#else /* defined(J9SHR_MSYNC_SUPPORT) */
+						*_runtimeFlags &= ~(J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT
 										| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT_ALL
 										| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT_RW
 										| J9SHR_RUNTIMEFLAG_MPROTECT_PARTIAL_PAGES_ON_STARTUP
 										| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT_PARTIAL_PAGES
 										| J9SHR_RUNTIMEFLAG_ENABLE_MPROTECT_ONFIND);
-#endif 
-					if (!_readOnlyOSCache) {
-						CC_TRACE(J9SHR_VERBOSEFLAG_ENABLE_VERBOSE, J9NLS_INFO, J9NLS_CC_PAGE_PROTECTION_UNSUPPORTED);
+						if (!_readOnlyOSCache) {
+							CC_TRACE(J9SHR_VERBOSEFLAG_ENABLE_VERBOSE, J9NLS_INFO, J9NLS_CC_PAGE_PROTECTION_UNSUPPORTED);
+						}
+#endif /* defined(J9SHR_MSYNC_SUPPORT) */
 					}
 				}
 
-#if defined (J9SHR_MSYNC_SUPPORT)
+#if defined(J9SHR_MSYNC_SUPPORT)
 				_doHeaderSync = _doReadWriteSync = _doSegmentSync = _doMetaSync = (*_runtimeFlags & J9SHR_RUNTIMEFLAG_ENABLE_MSYNC);
```

Comment:
This can move into the previous #if defined(J9SHR_MSYNC_SUPPORT)

Reply:
The previous `#if defined(J9SHR_MSYNC_SUPPORT)` is inside `if` condition.